Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -827,12 +833,152 @@ public class ProgramBuilder {
             return
         }
 
-        // Finally, insert the slice into the current program.
-        for instr in source {
-            if slice.contains(instr.index) {
-                adopt(instr, keepTypes: true)
+        
```

Comment:
Uff.. :'D Can you give an example for why this additional logic is required? I think that'd help.

Ideally this method would be as simple as possible since it's pretty important to the overall fuzzer performance and somewhat hard to automatically test for correctness.

Maybe we need a different algorithm to handle splicing instead of trying to make the old algorithm work under new constraints. For example, would it help to do multiple passes and precompute some bits of information first, for example block groups, dataflow dependencies, or active contexts?
I'll try to think of something once I better understand the issue this new logic is trying to handle.

Reply:
Here's an example that I've tried to simplify to explain the problem. Assume we have the following program that we want to splice at instruction `CallComputedMethod` which is at index 86, into a default context (.script):

```js
Splicing instruction 86 (CallComputedMethod) from 1F7EB535-2872-41AB-B2AF-170A9A2ED6FB
Instruction(op: Fuzzilli.LoadInteger, inouts_: [v0, v0])
...
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v6, v6])
Instruction(op: Fuzzilli.CreateArray, inouts_: [v6, v7, v7])
...
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v10, v10])
Instruction(op: Fuzzilli.CallFunction, inouts_: [v10, v11, v11])
...
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v26, v31])
...
Instruction(op: Fuzzilli.CreateArray, inouts_: [v0, v0, v31, v45])
Instruction(op: Fuzzilli.LoadInteger, inouts_: [v32, v46])
Instruction(op: Fuzzilli.StoreProperty, inouts_: [v31, v32, v47])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v33, v48])
Instruction(op: Fuzzilli.CallFunction, inouts_: [v33, v34, v49])
...
Instruction(op: Fuzzilli.BeginObjectDefinition, inouts_: [v38, v55])
Instruction(op: Fuzzilli.CreateSpreadProperty, inouts_: [v26, v56])
Instruction(op: Fuzzilli.CreateComputedProperty, inouts_: [v14, v32, v57])
Instruction(op: Fuzzilli.CreateProperty, inouts_: [v5, v58])
Instruction(op: Fuzzilli.BeginGetter, inouts_: [v59])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v39, v60])
Instruction(op: Fuzzilli.CallFunction, inouts_: [v39, v40, v61])
Instruction(op: Fuzzilli.EndGetterSetter, inouts_: [v62])
Instruction(op: Fuzzilli.BeginGetter, inouts_: [v63])
...
Instruction(op: Fuzzilli.CallMethod, inouts_: [v41, v31, v0, v10, v51, v72])
Instruction(op: Fuzzilli.EndGetterSetter, inouts_: [v73])
...
Instruction(op: Fuzzilli.BeginComputedSetter, inouts_: [v10, v53, v78])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v54, v79])
Instruction(op: Fuzzilli.LoadProperty, inouts_: [v54, v55, v80])
Instruction(op: Fuzzilli.StoreComputedProperty, inouts_: [v12, v55, v26, v81])
Instruction(op: Fuzzilli.LoadInteger, inouts_: [v56, v82])
Instruction(op: Fuzzilli.CreateArray, inouts_: [v56, v57, v83])
Instruction(op: Fuzzilli.StoreProperty, inouts_: [v57, v57, v84])
Instruction(op: Fuzzilli.LoadString, inouts_: [v58, v85])
Instruction(op: Fuzzilli.CallComputedMethod, inouts_: [v55, v58, v26, v7, v11, v33, v53, v6, v59, v86])
...
Instruction(op: Fuzzilli.EndGetterSetter, inouts_: [v89])
Instruction(op: Fuzzilli.EndObjectDefinition, inouts_: [v92])
```

With the existing splicing logic we end up with the following slice of instructions to be inserted:

```js
Instruction(op: Fuzzilli.LoadInteger, inouts_: [v0, v0])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v6, v6])
Instruction(op: Fuzzilli.CreateArray, inouts_: [v6, v7, v7])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v10, v10])
Instruction(op: Fuzzilli.CallFunction, inouts_: [v10, v11, v11])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v26, v31])
Instruction(op: Fuzzilli.CreateArray, inouts_: [v0, v0, v31, v45])
Instruction(op: Fuzzilli.LoadInteger, inouts_: [v32, v46])
Instruction(op: Fuzzilli.StoreProperty, inouts_: [v31, v32, v47])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v33, v48])
Instruction(op: Fuzzilli.BeginObjectDefinition, inouts_: [v38, v55])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v41, v64])
Instruction(op: Fuzzilli.CallMethod, inouts_: [v41, v31, v0, v10, v51, v72])
Instruction(op: Fuzzilli.BeginComputedSetter, inouts_: [v10, v53, v78])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v54, v79])
Instruction(op: Fuzzilli.LoadProperty, inouts_: [v54, v55, v80])
Instruction(op: Fuzzilli.LoadString, inouts_: [v58, v85])
Instruction(op: Fuzzilli.CallComputedMethod, inouts_: [v55, v58, v26, v7, v11, v33, v53, v6, v59, v86])
Instruction(op: Fuzzilli.EndGetterSetter, inouts_: [v89])
Instruction(op: Fuzzilli.EndObjectDefinition, inouts_: [v92])
```

Now if we were to adopt the instructions in sequential order, we would end up with syntactically invalid FuzzIL. This is because when we insert instruction `BeginObjectDef` it opens a restricted context and the instructions `LoadBuiltin` at index 64 and  and `CallMethod` at 72 are now going to be inserted into this restricted context (i.e. objectDefinition).

```
Instruction(op: Fuzzilli.BeginObjectDefinition, inouts_: [v38, v55])
Instruction(op: Fuzzilli.LoadBuiltin, inouts_: [v41, v64])
Instruction(op: Fuzzilli.CallMethod, inouts_: [v41, v31, v0, v10, v51, v72])
```

The reason these two instructions are captured in the slice is because they mutate an input which has a data dependency to one of the instructions included in the slice (in this instance `v10`).

To correctly handle these situations, we would need to hoist the defaulting instructions out of the restricted context and into a suitable instruction block. A suitable instruction block would be an open block (i.e we haven't processed its BlockEnd instruction yet) which opens and/or propagates a context that these instructions can be inserted into.

Reply:
Cool, that makes sense, thanks! Yeah I think we should try and come up with a simpler approach for computing slices, or this'll become unmaintainable :(

I think it could be possible to simplify things if we precompute for every instruction what the currently active context is (which we might need elsewhere, too, see other comment), and maybe also precompute all block instructions that belong together. Maybe we can further simplify things by completely ignoring blocks that don't surround the current slice (in your example above, the `BeginGetter` block), but I think it should work either way. If we have the context information for every instruction, we for example know whether we need to include a block begin instruction: if it opens a context that's required and doesn't exist at the instruction preceding it. Although we might anyway always want to select the closest block that creates a certain context, then it wouldn't really matter I think.

Another simplification might be to always include blocks entirely (with all instructions inside of them). This could make sense: if something depends e.g. on an argument to a class method, it probably doesn't really make sense to leave all other methods of the class empty. On the other hand, it might make slices significantly larger. I probably wouldn't mind that though if it kept the algorithm simpler. If we do that, things become a bit simpler: if we see an instruction with special context requirements (!= .script), then we know that we'll also include all instruction up to the next block that starts that context. Then we don't really need to push/pop entries onto/off our context stack until we find that opening block. Once we find it, we could "jump" to the end instruction (if we precomputed those), then include all instruction from there on up to the point where we started. Still without having to modify the context stack.

Finally, I think we should record more statistics about slices, as a start, maybe the total size of the slice and the relative size compared to the program it is taken from (`slice.count / program.count`).

But yeah, it's still all pretty vague. I'll try to come up with a proper algorithm. Let me know your thoughts!